crates/matrix-synapse/src/error.rs

```diff
@@ -0,0 +1,72 @@
+use std::{error::Error, fmt::Display};
+
+use http::Response;
+use mas_axum_utils::axum::body::Bytes;
+use serde::Deserialize;
+use tracing::debug;
+
+/// Represents a Matrix error
+/// Ref: <https://spec.matrix.org/v1.10/client-server-api/#standard-error-response>
+#[derive(Debug, Deserialize)]
+struct MatrixError {
+    errcode: String,
+    error: String,
+}
+
+/// Represents an error received from the homeserver.
+/// Where possible, we capture the Matrix error from the JSON response body.
+///
+/// Note that the `CatchHttpCodes` layer already captures the `StatusCode` for
+/// us; we don't need to do that twice.
+#[derive(Debug)]
+pub(crate) struct HomeserverError {
+    matrix_error: Option<MatrixError>,
+}
+
+impl Display for HomeserverError {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        if let Some(matrix_error) = &self.matrix_error {
+            write!(f, "{matrix_error}")
+        } else {
+            write!(f, "(no specific error)")
+        }
+    }
+}
+
+impl Error for HomeserverError {
+    fn description(&self) -> &str {
+        if let Some(matrix_error) = &self.matrix_error {
+            &matrix_error.error
+        } else {
+            "Homeserver Error"
+        }
+    }
```

Comment:
`description` has been deprecated for a while, it's delegated to Debug/Display now

Reply:
I think you still have to implement it? Maybe I should have just hardcoded a "don't description() me, you silly!" on it but ehh

Reply:
looks like the stdlib already does that for you :)

https://doc.rust-lang.org/stable/src/core/error.rs.html#110

Reply:
huh ok, weird that it added it (rust analyser?) for implementation then. Maybe it just always does that and I'm used to IntelliJ-Rust too much